[CuTe] Copy code from pycute for device mesh bookkeeping

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #161106
* #161016
* #162690
* #162414
* #162534
* __->__ #162413


We copied the whole module and its unit test into pytorch codebase. (https://github.com/NVIDIA/cutlass/blob/main/python%2Fpycute%2Flayout.py).

We did change the indentation of code from 2 spaces to 4 spaces. And add lint suppressor to make mypy happy.

Also we need to make changes to unit test to include ownership and use `run_tests, TestCase` so that the test gets picked up by CI.

cc @H-Huang @awgu @wanchaol @fegin @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim @dcci

fduwjj committed:
[CuTe] Copy code from pycute for device mesh bookkeeping

[ghstack-poisoned]

fduwjj committed:
Update on "[CuTe] Copy code from pycute for device mesh bookkeeping"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim

[ghstack-poisoned]

fduwjj commented:
@ezyang it is direct copy without any chance, the only change here is to make linter happy: add mypy suppression and formatting. I didn't copy all of them because we don't need. Also previous codes are in two different files, I just merge them into one.

ezyang commented:
Sorry where did you copy from? Here is what it looks like in CUTLASS: https://github.com/NVIDIA/cutlass/blob/main/python%2Fpycute%2Flayout.py

At minimum I was expecting to see 2 space indentation or you telling us you auto formatted it first. There's also three modules in the original.

ezyang commented:
We do need to reformat; I see now you have specified what changes you made, I am happy to trust you on those.

ezyang commented:
OK, so I see what you did is that you eliminated all code that wasn't being used by DeviceMesh. I guess this wasn't that difficult, so we can go this route.

ezyang commented:
(For future reference, something that I bet would have been quicker is to just import ALL the code without DCE'ing bits that aren't used; I would have also been OK with that.)

ezyang reviewed (approved):
You should also import in the tests (they're pretty short) and get them hooked up to CI. They will help us evaluate if we're doing refactors correctly.

fduwjj committed:
Update on "[CuTe] Copy code from pycute for device mesh bookkeeping"



This PR directly moved the code of `Layout`, `coalesce`, `composition`, `complement` and all necessary api into pytorch from [NVIDIA/cutlassmain/python%2Fpycute%2Flayout.py](https://github.com/NVIDIA/cutlass/blob/main/python%2Fpycute%2Flayout.py).

We did change the indentation of code from 2 spaces to 4 spaces. And add lint suppressor to make mypy happy.



cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

fduwjj committed:
Update on "[CuTe] Copy code from pycute for device mesh bookkeeping"



We copied the whole module and its unit test into pytorch codebase. (https://github.com/NVIDIA/cutlass/blob/main/python%2Fpycute%2Flayout.py).

We did change the indentation of code from 2 spaces to 4 spaces. And add lint suppressor to make mypy happy.



cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

ezyang commented on torch/distributed/pycute/__init__.py:
```diff
@@ -0,0 +1,38 @@
+# flake8: noqa
```
_pycute to be private?

fduwjj replied:
sure.

ezyang commented:
Looks good. BTW we probably don't need swizzle (and it has no deps from the other modules).

fduwjj committed:
Update on "[CuTe] Copy code from pycute for device mesh bookkeeping"



We copied the whole module and its unit test into pytorch codebase. (https://github.com/NVIDIA/cutlass/blob/main/python%2Fpycute%2Flayout.py).

We did change the indentation of code from 2 spaces to 4 spaces. And add lint suppressor to make mypy happy.



cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

fduwjj committed:
Update on "[CuTe] Copy code from pycute for device mesh bookkeeping"



We copied the whole module and its unit test into pytorch codebase. (https://github.com/NVIDIA/cutlass/blob/main/python%2Fpycute%2Flayout.py).

We did change the indentation of code from 2 spaces to 4 spaces. And add lint suppressor to make mypy happy.



cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

ezyang commented on test/distributed/_pycute/test_coalesce.py:
```diff
@@ -0,0 +1,99 @@
+# ruff: noqa: PGH004, G004, F403
```
Lol you should double check underscore in the test directory doesn't cause pytest to stop picking it up haha

fduwjj replied:
wait do you mean _pytest won't get picked up? But I do run into CI failure before I use "from torch.testing._internal.common_utils import run_tests, TestCase"
